src/transports/ipc.rs

```diff
@@ -0,0 +1,376 @@
+    self.thread.take()
+      .expect("We never touch thread except for drop; drop happens only once; qed")
+      .join()
+      .expect("Thread should shut down cleanly.");
+  }
+}
+
+/// Unix Domain Sockets (IPC) transport
+pub struct Ipc {
+  id: atomic::AtomicUsize,
+  write_sender: mpsc::Sender<String>,
+  read_sender: mpsc::Sender<(usize, Pending)>,
+}
+
+impl Ipc {
+  /// Create new IPC transport within existing Event Loop.
+  pub fn new<P>(path: P, handle: &reactor::Handle) -> Result<Self, IpcError> where
+    P: AsRef<Path>,
+  {
+    trace!("Connecting to: {:?}", path.as_ref());
+    let stream = UnixStream::connect(path, handle)?;
+    Self::with_stream(stream, handle)
+  }
+
+  /// Creates new IPC transport from existing `UnixStream` and `Handle`
+  fn with_stream(stream: UnixStream, handle: &reactor::Handle) -> Result<Self, IpcError> {
+    let (read, write) = stream.split();
+    let (write_sender, write_receiver) = mpsc::channel(2);
+    let (read_sender, read_receiver) = mpsc::channel(2);
+
+    let r = ReadStream {
+      read: read,
+      incoming: read_receiver,
+      pending: Default::default(),
+      buffer: vec![],
+      current_pos: 0,
+    };
+
+    let w = WriteStream {
+      write: write,
+      incoming: write_receiver,
+      state: WriteState::WaitingForRequest,
+    };
+
+    handle.spawn(r);
+    handle.spawn(w);
+
+    Ok(Ipc {
+      id: atomic::AtomicUsize::new(1),
+      write_sender: write_sender,
+      read_sender: read_sender,
+    })
+  }
+
+  /// Create new IPC transport with separate event loop.
+  /// NOTE: Dropping event loop handle will stop the transport layer!
+  pub fn with_event_loop<P>(path: P) -> Result<(EventLoopHandle, Self), IpcError> where
+    P: AsRef<Path>,
+  {
+    let done = Arc::new(atomic::AtomicBool::new(false));
+    let (tx, rx) = sync::mpsc::channel();
+    let path = path.as_ref().to_owned();
+    let done2 = done.clone();
+
+    let eloop = thread::spawn(move || {
+      let run = move || {
+        let event_loop = reactor::Core::new()?;
+        let ipc = Self::new(path, &event_loop.handle())?;
+        Ok((ipc, event_loop))
+      };
+
+      let send = move |result| {
+        tx.send(result).expect("Receiving end is always waiting.");
+      };
+
+      let res: Result<_, IpcError> = run();
+      match res {
+        Err(e) => send(Err(e)),
+        Ok((ipc, mut event_loop)) => {
+          send(Ok(ipc));
+
+          while !done2.load(atomic::Ordering::Relaxed) {
+            event_loop.turn(None);
+          }
+        },
+      }
+    });
+
+    rx.recv()
+      .expect("Thread is always spawned.")
+      .map(|ipc| (
+        EventLoopHandle { thread: Some(eloop), done: done },
+        ipc,
+      ))
+  }
+}
+
+impl Transport for Ipc {
+  type Out = BoxFuture<RpcValue, RpcError>;
+
+  fn execute(&self, method: &str, params: Vec<String>) -> Self::Out {
+    let id = self.id.fetch_add(1, atomic::Ordering::Relaxed);
```

Comment:
wouldn't it be safer to use an `AcqRel` or better yet `SeqCst` here? I think `fetch_add` is always interlocked on x86 but not sure about other platforms.